[feature] - RxJava2 - Kotlin extension: filterOrElse

neteinstein opened the pull request:
- `filterAndDo` - Kotlin extension to execute an action when a stream is filtered.
- Added commit template for this project

neteinstein committed:
[feature] RxJava2 - Call a function when a stream is filtered

Kotlin extension to execute an action when a stream is filtered.

neteinstein committed:
[defaults] Added commit template

Added default commit template. Please add default config via
git config --get commit.template or  git config --local commit.template

neteinstein committed:
[fix] Observable filterAndDo

Only the first event on the stream would execute the pre
validation to show the dialog, now fixed.

neteinstein committed:
[feature] RxJava2 - Call a function when a stream is skipped

Kotlin extension to execute an action when a stream is skipped.

neteinstein committed:
[fix] Single filterAndDo

Return a Single instead of a Maybe

joaocaseiro commented on rxjava/src/main/kotlin/com/mindera/skeletoid/rxjava/Observable.kt:
```diff
@@ -38,29 +45,50 @@ fun <T : Any> Observable<T>.createUniqueConcurrentRequestCache(requestMap: Concu
+            elseFunction()
+        }
+    }
+        .filter { condition }
```
can't we do:
this.filter { condition }.doOnNext { elseFunction() }
?

neteinstein replied:
The parameter name made you think it's a different thing. It not to be run as else, but when the condition is met.
So it should be before or it won't run. I've changed the name to `functionToRunIfConditionIsMet`

joaocaseiro replied:
@neteinstein I didn't even noticed the parameter name (really else function is the exact opposite of what it is doing).
When you do .filter { condition }, it will only "let through" the inputs when the condition is true, so it will should be exactly as if there was an if(condition) and only continue the chain if the condition is true. The resulting observable would only then be composed of the inputs where the condition is true. 
So doing onNext in that Observable is the same as onNext with an if condition in the unfiltered observable, without checking the condition twice.

joaocaseiro replied:
Test 1:
        Observable.just(1,2,3,4,5,6,7,8,9,10).filter { it % 2 == 0 }.doOnNext { Log.v(TAG, "Test 2 Number $it is pair") }.subscribe {
            Log.v(TAG, "Test 2 Subscribe $it")
        }.addTo(disposables)
Output:
Test Number 2 is pair
Test Subscribe 2
Test Number 4 is pair
Test Subscribe 4
Test Number 6 is pair
Test Subscribe 6
Test Number 8 is pair
Test Subscribe 8
Test Number 10 is pair
Test Subscribe 10

Test 2:
Observable.just(1,2,3,4,5,6,7,8,9,10).doOnNext {
            if(it % 2 == 0) {
                Log.v(TAG, "Test 1 Number $it is pair")
            }
        }.filter { it % 2 == 0 }.subscribe {
            Log.v(TAG, "Test 1 Subscribe $it")
        }.addTo(disposables)
Output:
Test 1 Number 2 is pair
Test 1 Subscribe 2
Test 1 Number 4 is pair
Test 1 Subscribe 4
Test 1 Number 6 is pair
Test 1 Subscribe 6
Test 1 Number 8 is pair
Test 1 Subscribe 8
Test 1 Number 10 is pair
Test 1 Subscribe 10

bruno-garces replied:
@neteinstein you probably meant this
```
doOnNext { if (!condition) elseFunction() }
.filter { condition }
```

neteinstein replied:
Yep @bruno-garces . Just got confused 😅.
will revert to that.

neteinstein replied:
Please validate now @Jmvcaseiro @bruno-garces

joaocaseiro replied:
I commented in a new place

djrsousa committed:
[fix] Fixes Single filterAndDo

neteinstein committed:
[fix] Changed filterAndDo parameter name

Parameter called elseFunction but should be called functionToRunIfConditionIsMet

neteinstein committed:
Revert "[fix] Changed filterAndDo parameter name"

This reverts commit df459cd1e91a47f664f62b6cbc1bb6004a7c2176.

neteinstein committed:
[fix] WGA-4307 - Changed filterAndDo parameter name

Parameter called elseFunction but should be called runIfFiltered

joaocaseiro commented on rxjava/src/main/kotlin/com/mindera/skeletoid/rxjava/Observable.kt:
```diff
@@ -46,21 +53,40 @@ internal data class DataHolder<T>(val something: T? = null, val throwable: Throw
+        if (condition) {
+            runIfFiltered()
+        }
+    }.filter { condition }
```
I am very confused at this point, but if you want to have this code structure instead of my suggestion you should do two things:
- change the name to something like "filterOrDo"
- change the if to something like "if(!condition)"
You may be thinking that .filter(...) will filter when you return true but it's the other way around (check my code and its output in the previous comment about this)

neteinstein replied:
change the if to something like "if(!condition)" -> It's done, check the latest version of the code

djrsousa commented on rxjava/src/main/kotlin/com/mindera/skeletoid/rxjava/Maybe.kt:
```diff
@@ -22,4 +22,12 @@ fun <T : Any> Maybe<T>.subscribeOnMain(): Maybe<T>
\ No newline at end of file
+        = observeOn(AndroidSchedulers.mainThread())
+
+fun <T> Maybe<T>.filterAndDo(condition: Boolean, runIfFiltered: () -> Unit): Maybe<T> =
```
I'd prefer to name it `filterOrElse` and instead of `runIfFiltered` I'd just call it `action`. What do you think?

neteinstein committed:
[fix] Changed filterAndDo to filterOrElse and skipWhileAndDo

Fix method names according to their logic

djrsousa committed:
[fix] Uniform generic extensions functions.

djrsousa committed:
[fix] Adds ProxyUtils

djrsousa committed:
[fix] Code changes as per code review feedback.

djrsousa committed:
[fix] Code format.

bruno-garces commented on rxjava/src/main/kotlin/com/mindera/skeletoid/rxjava/Maybe.kt:
```diff
@@ -22,4 +22,11 @@ fun <T : Any> Maybe<T>.subscribeOnMain(): Maybe<T>
\ No newline at end of file
+        = observeOn(AndroidSchedulers.mainThread())
+
+fun <T : Any> Maybe<T>.filterOrElse(condition: Boolean, action: () -> Unit): Maybe<T> =
```
We should add some tests 😐

joaocaseiro replied:
orElse...

djrsousa replied:
Done.

djrsousa committed:
[fix] Adds filterOrElse unit tests.

djrsousa committed:
[fix] Adds skipWhileAndDo unit tests.
